### ISSUE TITLE
TempDirectory breaks constructor parameter resolution

### ISSUE THREAD

**Issue by sbrannen** [junit-team/junit5#1748]
## Overview

Changes made in #1746 prevent test classes from having any parameters in their constructor if the `TempDirectory` extension is registered.

## Failing Example

The following test class fails against JUnit Jupiter `5.4.0-RC1` simply because the `TempDirectory` extension is registered.

```java
@ExtendWith(TempDirectory.class)
static class BrokenTestCase {

	BrokenTestCase(TestInfo testInfo) {
	}

	@Test
	void test() {
	}

}
```

## Related Issues

- #1746
- #1747

## Deliverables

- [x] Ensure that the `TempDirectory` extension does not prevent the use of constructor arguments.
- [x] Consider documenting the fix in the Release Notes.

**Comment by sbrannen** [comment]
Reopening this issue to _consider documenting the fix in the Release Notes_.

**Comment by sormuras** [comment]
> In other words, if a developer chooses per-class lifecycle semantics for a given test class, I don't see any problem with using @TempDir on a constructor parameter.

Seems reasonable. +1

**Comment by snicoll** [comment]
I was very confused when I first used this feature (from the pioneer project). For the record, I very much prefer preventing the use of such extension in the constructor rather than allowing it and passing the same value for each test invocation. 

My preference, of course, would be to allow constructor injection and inject a different value every time the class is instantiated.

**Comment by filiphr** [comment]
Regarding the constructor parameter, when I initially used the `ParameterResolver` I thought that it would be possible to inject into the constructor. Isn't it possible to use that or maybe a new `ConstructorParameterResolver` that would allow passing parameters to the constructor?

**Comment by sbrannen** [comment]
> Regarding the constructor parameter, when I initially used the `ParameterResolver` I thought that it would be possible to inject into the constructor. Isn't it possible to use that or maybe a new `ConstructorParameterResolver` that would allow passing parameters to the constructor?

Parameter resolution for constructors via the `ParameterResolver` extension API is not a problem in and of itself.

For example, that works fine with things like `TestInfo` and `TestReporter`. It also works fine for frameworks like the Spring Framework, since Spring looks up the component in the current `ApplicationContext`.

Parameter resolution for constructors only becomes "unintuitive" (i.e., has undesirable side effects) if the resolved argument is stored in the `ExtensionContext.Store`, which is what an extension like `TempDirectory` is basically forced to do.

**Comment by sbrannen** [comment]
> I was very confused when I first used this feature (from the pioneer project). 

I can fully understand the confusion.

It's _confusing_ for the core members of the JUnit 5 team as well.

> For the record, I very much prefer preventing the use of such extension in the constructor rather than allowing it and passing the same value for each test invocation.

Thanks for the feedback on that point. That's also our preference and thus the rationale behind disallowing `@TempDir` on constructor parameters for the time being (though potentially re-allowing them if the test class is configured with test instance per-class lifecycle semantics).

> My preference, of course, would be to allow constructor injection and inject a different value every time the class is instantiated.

We (the JUnit 5 team) would also prefer that. It makes the most sense in terms of user expectations.

I actually prototyped a bit locally to see if I could come up with a solution along those lines.

Unfortunately, it is by no means easy to achieve that with the current implementation of the internals of JUnit Jupiter (with regard to scoping of `ExtensionContext` types). Basically, the `ClassExtensionContext` (which is passed to a `ParameterResolver`) does not have access to the `ExtensionContext.Store` that is used by the `MethodExtensionContext`. Consequently, a `ParameterResolver` has no reliable way of providing a temp directory that is _scoped_ for the current test method.

One can argue that this is a flaw in the design/implementation of JUnit Jupiter. In any case, that is the status quo, and it's too late to make drastic changes along those lines in 5.4. So we'll have to see if we can make any improvements on this front in the 5.5+ timeline.

For the time being, field injection is likely the best option -- unless we decide to allow `@TempDir` on constructor parameters for _per-class_ lifecycle semantics prior to the 5.4 GA release.

**Comment by sbrannen** [comment]
**Proposal**:

Proposed change to `TempDirectory`.

```java
public boolean supportsParameter(ParameterContext parameterContext, ExtensionContext extensionContext) {
	boolean annotated = parameterContext.isAnnotated(TempDir.class);

	if (annotated) {
		boolean inConstructor = parameterContext.getDeclaringExecutable() instanceof Constructor;
		boolean testInstancePerMethod = extensionContext.getTestInstanceLifecycle()//
				.orElse(null) == Lifecycle.PER_METHOD;

		if (inConstructor && testInstancePerMethod) {
			throw new ParameterResolutionException(
				"@TempDir is not supported on constructor parameters when using @TestInstance(PER_METHOD) semantics. "
						+ "Please use field injection instead.");
		}
	}

	return annotated;
}
```

**Comment by mmerdes** [comment]
> @junit-team/junit-lambda,
> 
> I've added the `team discussion` label to discuss the following in conjunction with this issue as well.
> 
> Do we want to completely forbid the use of `@TempDir` on constructor parameters?
> 
> Or... should we only forbid that for test instance _per-method_ lifecycle semantics?
> 
> In other words, if a developer chooses _per-class_ lifecycle semantics for a given test class, I don't see any problem with using `@TempDir` on a constructor parameter.
> 
> Thoughts?

In the interest of simplicity and consisteny I would prefer a complete ban of `@TempDir` on constructor parameters rather than the 'mixed' solution.

**Comment by marcphilipp** [comment]
I tend to agree with @mmerdes. If there's a compelling use case for constructor injection in conjunction with the per-class lifecycle we could add support at a later point in time.

**Comment by sbrannen** [comment]
@sormuras, @mmerdes, & @marcphilipp, thanks for the feedback.

That's basically a 50/50 draw, and I agree that we can add that support later without it being a breaking change. 

So let's leave it _as is_ for now.

**Comment by sbrannen** [comment]
I am no closing this issue.

Documentation in the release notes will be addressed in #1749.
